server/src/test/java/org/elasticsearch/cluster/routing/allocation/allocator/AllocationActionListenerTests.java

```diff
@@ -59,4 +66,36 @@ public void testShouldDelegateFailureImmediately() {
 
         assertThat(completed.get(), equalTo(true));
     }
+
+    public void testShouldExecuteWithCorrectContext() throws Exception {
+
+        var queue = new DeterministicTaskQueue();
+        var pool = queue.getThreadPool();
+        pool.getThreadContext().addResponseHeader("header", "root");
+
+        var completed = new AtomicReference<String>();
+        var listener = new AllocationActionListener<AcknowledgedResponse>(
+            ActionListener.wrap(
+                ignore -> completed.set(pool.getThreadContext().getResponseHeaders().get("header").get(0)),
+                exception -> { throw new AssertionError("Should not fail in test"); }
+            ),
+            pool.getThreadContext()
+        );
+
+        pool.generic().execute(() -> {
+            pool.getThreadContext().addResponseHeader("header", "clusterStateUpdate");
+            listener.clusterStateUpdate().onResponse(AcknowledgedResponse.TRUE);
+        });
+        pool.generic().execute(() -> {
+            pool.getThreadContext().addResponseHeader("header", "reroute");
+            listener.reroute().onResponse(null);
+        });
+
+        queue.runAllTasks();
+        assertThat(completed.get(), equalTo("root"));
+    }
```

Comment:
I wonder if we need the thread pool here? How about this?

```suggestion
    public void testShouldExecuteWithCorrectContext() {

        var threadContext = new ThreadContext(Settings.EMPTY);
        var completed = new AtomicReference<String>();
        AllocationActionListener<AcknowledgedResponse> listener;

        threadContext.putHeader("header", "root");
        listener = new AllocationActionListener<>(
            ActionListener.wrap(
                ignore -> completed.set(threadContext.getHeader("header")),
                exception -> {
                    throw new AssertionError("Should not fail in test");
                }
            ),
            threadContext
        );

        try (var ignored = threadContext.stashContext()) {
            threadContext.putHeader("header", "clusterStateUpdate");
            listener.clusterStateUpdate().onResponse(AcknowledgedResponse.TRUE);
        }

        try (var ignored = threadContext.stashContext()) {
            threadContext.putHeader("header", "reroute");
            listener.reroute().onResponse(null);
        }

        assertThat(completed.get(), equalTo("root"));
    }
```

NB using request headers in this test - response headers are not cleared in a `ContextPreservingActionListener`, and in fact you get all three response headers in the result so the following listener will fail.

```
            ActionListener.wrap(
                ignore -> {
                    final var testHeaders = pool.getThreadContext().getResponseHeaders().get("header");
                    assertThat(testHeaders, hasSize(1));
                    completed.set(testHeaders.get(0));
                },
                exception -> { throw new AssertionError("Should not fail in test"); }
            ),
```

Alternatively we could assert that we do want this response-header-preserving behaviour here.